processing/src/main/java/io/druid/query/aggregation/FilteredAggregatorFactory.java

```diff
@@ -140,6 +144,61 @@ public int getMaxIntermediateSize()
     return delegate.getMaxIntermediateSize();
   }
 
+  @Override
+  public AggregatorFactory optimizeForSegment(PerSegmentQueryOptimizationContext optimizationContext)
+  {
+    if (filter instanceof IntervalDimFilter) {
```

Comment:
It looks that this optimization would work only when there is a single `IntervalDimFilter` here. But, probably there would be more complicated filter using `AndDimFilter` or `OrDimFilter`. To support that, we need to check all filters. What do you think about adding a new interface like `optimizeForSegment` to `DimFilter`? Probably we also need to use the visitor pattern to fully optimize this.

I'm not sure this should be a part of this PR. If you think it's not, would you please raise an issue about this and add a link here?

Reply:
Hm, I think it makes sense to do similar optimizations when possible with more complex AND/OR + Interval filters, I was planning on addressing just the simplest case in this PR, I'll open an issue for that.

Re: adding a visitor pattern on the filters, I considered doing that, but decided to stop at the aggregator level for this PR. (the goal in my mind for this PR was to support a specific time-based optimization for IntervalDimFilter and add general interfaces/methods to allow queries/aggs to optimize themselves on a per-segment basis).

I think I'll do that in a follow up PR that adds support for optimizing on AND/OR filters as well, the interface I think would look something like:

```
PerSegmentFilterOptimizeResult optimizeFilterForSingleSegmentFilteredAggregator(PerSegmentOptimizationContext optimizationContext)

class PerSegmentFilterOptimizeResult {
 DimFilter -> optimized filter
 OptimizeStatus -> enum of MATCHES_ALL, PARTIAL_MATCH, MATCHES_NONE, determines whether FilteredAggregator removes the filter, keeps the filter, or suppresses aggregations
}
```

Reply:
Sounds nice!